apps/prairielearn/src/question-servers/freeform.js

```diff
@@ -1134,20 +1134,22 @@ async function renderPanel(panel, codeCaller, variant, submission, course, local
   }
 
   const data = {
-    params: _.get(variant, 'params', {}),
-    correct_answers: _.get(variant, 'true_answer', {}),
-    submitted_answers: submission ? _.get(submission, 'submitted_answer', {}) : {},
+    // `params` and `true_answer` are allowed to change during `parse()`/`grade()`,
+    // so we'll use the submission's values if they exist.
+    params: submission?.params ?? variant.params ?? {},
+    correct_answers: submission?.true_answer ?? variant.true_answer ?? {},
```

Comment:
@jonatanschroeder I talked about this with Matt, and we think this is a reasonable middle-ground behavior for now. Rather than rendering panels with the variant's `params`/`correct_answers`, we'll do so with the one stored on the submission.

We decided to not yet make the change where we don't update `variants.{params,true_answer}` columns once the variant has been created. We _already_ update those columns on the variants if `grade()` changes them, so doing the same with `parse()` is at least consistent with other existing behavior. Immutable `variants` columns is still an interesting idea, but that potentially has other implications so we don't want to do that just yet.

Reply:
Sounds good. There are a few other pieces of code that also use variant values instead of submission values, up to you if you want to update them or not, but I think they are in a similar level as rendering:
https://github.com/PrairieLearn/PrairieLearn/blob/ee607c24e6461abe4abd4edb6a017292f9461da1/apps/prairielearn/src/lib/externalGraderCommon.ts#L83-L87
https://github.com/PrairieLearn/PrairieLearn/blob/ee607c24e6461abe4abd4edb6a017292f9461da1/apps/prairielearn/src/lib/manualGrading.ts#L138-L143
https://github.com/PrairieLearn/PrairieLearn/blob/ee607c24e6461abe4abd4edb6a017292f9461da1/apps/prairielearn/src/pages/instructorAssessmentManualGrading/instanceQuestion/rubricSettingsModal.html.ts#L402-L413

Also, assessment downloads may be in the same boat.

Reply:
I changed this in 4895cb9, and I found a few other places where this could be changed too. Note that because we always write the latest values back to the variant anyways, this shouldn't actually result in a change in behavior. It does however feel more correct, and it does set us up to potentially make the variant params immutable in the future.

I didn't change the rubric settings modal, as I don't think we're guaranteed to have a submission there (and at any rate, if a submission is present there, it should always match the data from the variant).

Reply:
> I didn't change the rubric settings modal, as I don't think we're guaranteed to have a submission there (and at any rate, if a submission is present there, it should always match the data from the variant).

We are guaranteed to have a submission present there:
https://github.com/PrairieLearn/PrairieLearn/blob/ee607c24e6461abe4abd4edb6a017292f9461da1/apps/prairielearn/src/pages/instructorAssessmentManualGrading/instanceQuestion/instanceQuestion.ts#L24-L37

That said, I understand your decision. Also, in the rubric setting we're only using the keys, so this shouldn't make a big difference even if we end up making the variant values immutable.

Reply:
Ah, interesting - the `resLocals.submission?.submitted_answer` optional chaining threw me off! I'll still keep the code as is.